lib/run-compilation.js

```diff
@@ -46,8 +41,15 @@ module.exports = function runCompilation(options, compilation, done) {
     });
 
   // eslint-disable-next-line no-unused-expressions
-  compilation.plugin && compilation.plugin('compilation', function (compilation) {
-    compilation.errors = compilation.errors.concat(errors);
-    compilation.warnings = compilation.warnings.concat(warnings);
+  compiler.plugin('compilation', function (compilation) {
```

Comment:
This is done.

I wasn't sure why this guard was there - what race condition do you mean? In my tests in watch mode, it was working correctly.